File: mobile/modules/bluetooth-sdk/android/src/main/java/com/mentra/bluetoothsdk/streaming/StreamKeepAliveAckWindow.kt
package com.mentra.bluetoothsdk

import java.util.ArrayDeque

/**
 * Tracks a bounded, ordered window of stream keep-alive requests.
 *
 * ACKs can arrive after a newer request has already been sent. Matching any request still in the
 * window proves that the stream transport is responsive; requests older than the match can then be
 * discarded as stale.
 */
internal class StreamKeepAliveAckWindow(
    private val maxTrackedAckIds: Int,
    private val maxMissedAcks: Int,
) {
    private val pendingAckIds = ArrayDeque<String>()
    var missedAckCount: Int = 0
        private set
    var armed: Boolean = false
        private set
    private var didReportTimeout = false

    init {
        require(maxTrackedAckIds > 0)
        require(maxMissedAcks > 0)
    }

    fun arm() {
        armed = true
        resetPendingState()
    }

    fun recordSent(ackId: String) {
        pendingAckIds.remove(ackId)
        pendingAckIds.addLast(ackId)
        while (pendingAckIds.size > maxTrackedAckIds) {
            pendingAckIds.removeFirst()
        }
    }

    /** Returns the miss count once, when the reporting threshold is first reached. */
    fun recordTick(): Int? {
        if (!armed || pendingAckIds.isEmpty()) return null
        missedAckCount += 1
        if (missedAckCount < maxMissedAcks || didReportTimeout) return null
        didReportTimeout = true
        return missedAckCount
    }

    fun acknowledge(ackId: String): Boolean {
        if (!pendingAckIds.contains(ackId)) return false
        while (pendingAckIds.isNotEmpty()) {
            if (pendingAckIds.removeFirst() == ackId) break
        }
        missedAckCount = 0
        didReportTimeout = false
        return true
    }

    private fun resetPendingState() {
        pendingAckIds.clear()
        missedAckCount = 0
        didReportTimeout = false
    }
}

File: mobile/modules/bluetooth-sdk/ios/Source/streaming/StreamKeepAliveAckWindow.swift
import Foundation

/// Tracks a bounded, ordered window of stream keep-alive requests.
///
/// ACKs can arrive after a newer request has already been sent. Matching any
/// request still in the window proves that the stream transport is responsive;
/// requests older than the match can then be discarded as stale.
final class StreamKeepAliveAckWindow {
    private let maxTrackedAckIds: Int
    private let maxMissedAcks: Int
    private var pendingAckIds: [String] = []
    private(set) var missedAckCount = 0
    private(set) var armed = false
    private var didReportTimeout = false

    init(maxTrackedAckIds: Int, maxMissedAcks: Int) {
        precondition(maxTrackedAckIds > 0)
        precondition(maxMissedAcks > 0)
        self.maxTrackedAckIds = maxTrackedAckIds
        self.maxMissedAcks = maxMissedAcks
    }

    static func forStartedStream(maxTrackedAckIds: Int, maxMissedAcks: Int) -> StreamKeepAliveAckWindow {
        let window = StreamKeepAliveAckWindow(
            maxTrackedAckIds: maxTrackedAckIds,
            maxMissedAcks: maxMissedAcks
        )
        window.arm()
        return window
    }

    func arm() {
        armed = true
        resetPendingState()
    }

    func recordSent(ackId: String) {
        pendingAckIds.removeAll { $0 == ackId }
        pendingAckIds.append(ackId)
        if pendingAckIds.count > maxTrackedAckIds {
            pendingAckIds.removeFirst(pendingAckIds.count - maxTrackedAckIds)
        }
    }

    /// Records a heartbeat interval without a matching ACK. Returns the miss
    /// count once, when the reporting threshold is first reached.
    func recordTick() -> Int? {
        guard armed, !pendingAckIds.isEmpty else { return nil }
        missedAckCount += 1
        guard missedAckCount >= maxMissedAcks, !didReportTimeout else { return nil }
        didReportTimeout = true
        return missedAckCount
    }

    func acknowledge(ackId: String) -> Bool {
        guard let matchIndex = pendingAckIds.firstIndex(of: ackId) else { return false }
        pendingAckIds.removeFirst(matchIndex + 1)
        missedAckCount = 0
        didReportTimeout = false
        return true
    }

    private func resetPendingState() {
        pendingAckIds.removeAll(keepingCapacity: true)
        missedAckCount = 0
        didReportTimeout = false
    }
}
